tools/sctrace/README.md

```diff
@@ -0,0 +1,357 @@
+
+```scml
+poll(fds = [{ fd, events = POLLIN | POLLOUT, .. }], nfds, timeout);
+```
+
+## Examples
+
+### Example 1: Basic File Operations
+
+Create `file_ops.scml`:
+```scml
+openat(dirfd, flags = O_RDONLY | O_WRONLY | O_RDWR, mode);
+read(fd, buf, count = <INTEGER>);
+write(fd, buf, count = <INTEGER>);
+close(fd);
+```
+
+Run:
+```bash
+sctrace file_ops.scml -- cat /etc/passwd
+```
+
+### Example 2: Network Operations
+
+Create `network.scml`:
+```scml
+socket(domain = AF_INET | AF_INET6, type = SOCK_STREAM | SOCK_DGRAM, protocol);
+connect(sockfd, addr, addrlen);
+send(sockfd, buf, len, flags);
+recv(sockfd, buf, len, flags);
+```
+
+Run:
+```bash
+sctrace network.scml -- curl http://example.com
+```
+
+### Example 3: Using Asterinas Compatibility Patterns
+
+Use the provided [asterinas.scml](../../book/src/kernel/linux-compatibility/limitations-on-system-calls/asterinas.scml) (work in progress) and
+test with various commands:
+
+```bash
+# Monitor file system operations
+sctrace asterinas.scml -- tree .
+
+# Monitor process information calls
+sctrace asterinas.scml -- top
+
+# Monitor network operations
+sctrace asterinas.scml -- ping 127.0.0.1
+```
+
+### Example 4: Offline Analysis
+
+```bash
+# Generate trace log
+strace -o trace.log ls -la
+
+# Analyze with sctrace
+sctrace patterns.scml --input trace.log
+```
+
+## strace Output Format Support
+
+sctrace supports parsing various strace output formats, including multi-threaded program traces.
+
+### Single-threaded Output
+
+Standard strace output for single-threaded programs:
+```
+openat(AT_FDCWD, "/etc/passwd", O_RDONLY) = 3
+read(3, "root:x:0:0:root:/root:/bin/bash\n"..., 4096) = 1234
+close(3) = 0
+```
+
+### Multi-threaded Output
+
+When tracing multi-threaded programs, strace may split system calls across multiple lines due to thread interleaving. sctrace automatically handles this reconstruction:
+
+**Blocked syscall format:**
+```
+1234 openat(AT_FDCWD, "/path/to/file", O_RDONLY <unfinished ...>
+```
+
+**Resumed syscall format:**
+```
+1234 <... openat resumed>) = 3
+```
+
+**Automatic reconstruction:**
+sctrace internally reconstructs these into complete syscalls:
+```
+1234 openat(AT_FDCWD, "/path/to/file", O_RDONLY) = 3
+```
+
+### Signal and Exit Lines
+
+sctrace automatically skips signal and process exit information:
+
+**Signal lines (skipped):**
+```
+--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=1234, ...} ---
+```
+
+**Exit status lines (skipped):**
+```
++++ exited with 0 +++
+```
+
+## Output
+
+sctrace provides colored output to distinguish between supported and unsupported system calls:
+
+- **Supported calls**: Normal output (or hidden in quiet mode)
+- **Unsupported calls**: Highlighted in red with "unsupported" message
+
+### Example Output
+
+```
+openat(AT_FDCWD, "/etc/passwd", O_RDONLY) = 3
+read(3, "root:x:0:0:root:/root:/bin/bash\n"..., 4096) = 1234
+close(3) = 0
+chmod("/tmp/test", 0755) (unsupported)
+```
+
+## Project Structure
```

Comment:
Remove this section. There are less than 10 files in this project. The code organization is quite straightforward.